test/lint/lint-filenames-permissions.py

```diff
@@ -0,0 +1,182 @@
+            # allowable set of shebangs
+            if file_extension in ALLOWED_EXECUTABLE_SHEBANG.keys():
+                if shebang not in ALLOWED_EXECUTABLE_SHEBANG[file_extension]:
+                    print(
+                        f"""File "{filename}" is missing expected shebang """
+                        + " or ".join(
+                            [x.decode("utf-8") for x in ALLOWED_EXECUTABLE_SHEBANG[file_extension]]
+                        )
+                    )
+                    exit_code = 1
+
+        elif file_perms == ALLOWED_PERMISSION_NON_EXECUTABLES:
+            continue
+        else:
+            print(
+                f"""File "{filename}" has unexpected permission {file_perms}. Do "chmod {ALLOWED_PERMISSION_NON_EXECUTABLES} {filename}" (if non-executable) or "chmod {ALLOWED_PERMISSION_EXECUTABLES} {filename}" (if executable)."""
+            )
+            exit_code = 1
+
+    return exit_code
+
+
+def check_shebang_file_permissions() -> int:
+    """
+    Checks every file that contains a shebang line to ensure it has an executable permission
+    """
+    filenames = check_output(CMD_SHEBANG_FILES, shell=True).decode("utf8").strip().split("\n")
+
+    # The git grep command we use returns files which contain a shebang on any line within the file
+    # so we need to filter the list to only files with the shebang on the first line
+    filenames = [filename.split(":1:")[0] for filename in filenames if ":1:" in filename]
+
+    exit_code = 0
+    for filename in filenames:
+        file_extension = get_filename_extension(filename)
+        file_perms = int(oct(os.stat(filename).st_mode)[-3:])
+
+        if file_perms != ALLOWED_PERMISSION_EXECUTABLES:
+            # These file types are typically expected to be sourced and not executed directly
+            if file_extension in ["bash", "init", "openrc", "sh.in"]:
+                continue
+
+            # *.py files which don't contain an `if __name__ == '__main__'` are not expected to be executed directly
+            if file_extension == "py":
+                file_data = open(filename, "r", encoding="utf8").read()
+                if not re.search("""if __name__ == ['"]__main__['"]:""", file_data):
+                    continue
+
+            print(
+                f"""File "{filename}" contains a shebang line, but has the file permission {file_perms} instead of the expected executable permission {ALLOWED_PERMISSION_EXECUTABLES}. Do "chmod {ALLOWED_PERMISSION_EXECUTABLES} {filename}" (or remove the shebang line)."""
+            )
+            exit_code = 1
+    return exit_code
+
+
+def main() -> NoReturn:
+    failed_tests = 0
+    failed_tests += check_all_filenames()
+    failed_tests += check_source_filenames()
+    failed_tests += check_all_file_permissions()
+    failed_tests += check_shebang_file_permissions()
+
+    if failed_tests:
+        sys.exit(1)
```

Comment:
Nit: Given that `failed_tests` number is computed, one can probably print the number of failed tests too.

Reply:
Great idea, I've updated to keep track of the number of failures, and print a message with the total number of failures, if any :)